packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js

```diff
@@ -1376,6 +1376,64 @@ const tests = {
         }
       `,
     },
+    {
+      code: normalizeIndent`
+        function useFoo(foo){
+          return useMemo(() => foo, [foo]);
+        }
+      `,
+    },
+    {
+      code: normalizeIndent`
+        function useFoo(){
+          const foo = "hi!";
+          return useMemo(() => foo, [foo]); 
+        }
+      `,
+    },
+    {
+      code: normalizeIndent`
+        function useFoo(){
+          let foo = {};
+          foo = 1;
+          return useMemo(() => foo, [foo]);
+        }
+      `,
+    },
+    {
+      code: normalizeIndent`
+        function useFoo(){
+          let {foo} = {foo: 1};
+          return useMemo(() => foo, [foo]);
+        }
+      `,
+    },
+    {
+      code: normalizeIndent`
+        function useFoo(){
+          let [foo] = [1];
+          return useMemo(() => foo, [foo]);
+        }
+      `,
+    },
+    {
+      code: normalizeIndent`
+        function useFoo() {
+          const foo = "fine";
+          if(true) {
+            const foo = {};
+          }
+          return useMemo(() => foo, [foo]);
```

Comment:
Add a comment for why this doesn't warn. (A limitation.) It's not obvious.

Reply:
Note that this isn't a limitation - it's just not an error. The inner `const foo = {}` is a separate variable in a separate child scope, so its `{}` value is unrelated to the `const foo = "fine"` in the component's scope.

Though a comment explaining that would be worthwhile!

Reply:
Ahh.

Reply:
Added a test in https://github.com/facebook/react/pull/19590/commits/eaa2aff9424e7ded164799fa2aef27ccfa4ed3d0